### ISSUE TITLE
`clang_delta`: `query-instances` with `getQualifiedNameAsString`

### ISSUE THREAD

**Issue by andrewvaughanj** [marxin/cvise#43]
Sorry if this is a "dumb" question, but what's the status of making changes to `clang_delta` in the `cvise` repo? I am not sure how closely this repo tracks `creduce`, or if `cvise` is happy to have feature extensions to `clang_delta`.

Why am I asking? Well, I am interested (in some circumstances) in running `clang_delta` to get the instance number *as well as the function name* for a given transformation.

For some inputs, I know _a priori_ a function I want to _preserve_, but I want to replace every other function with the declaration. If `query-instances` could not only give the number of instances, but also the function _name_ for each instance, that would be awesome.

If I had this, I could do something like:

```
# query all of the instances in our original file
clang_delta --query-instances=replace-function-def-with-decl test.cpp

# record the index of the function I care about; let's say it is index 3

# replace from the first to the second (leaving the third alone!)
clang_delta --transformation=replace-function-def-with-decl --counter=1 --to-counter=2 test.cpp > intermediate.cpp

# re-query to get the remaining numbers
clang_delta --query-instances=replace-function-def-with-decl intermediate.cpp

# validate that the original function should now be index 1

# replace from the second to the total number
clang_delta --transformation=replace-function-def-with-decl --counter=2 --to-counter=99 intermediate.cpp > final.cpp

# the file final.cpp now *only* has a body for the function that was at index 3 in the original file
```

I've taken a look at this, and it seems that it is achievable by using `getQualifiedNameAsString` inside of `addOneFunctionDef`.

If I was to wrap this up as a feature request, is it something that `cvise` would be happy to accept?

**Comment by andrewvaughanj** [comment]
A (very) rough implementation of what I'm talking about is here:

- https://github.com/marxin/cvise/compare/master...andrewvaughanj:dump_transformation_instances?expand=1

(a number of these changes are due to `git-clang-format`; apologies if these are unwarranted)

Output looks like this:

```
avj@stoat:~/cvise> cat test.cpp
class Moo
{
    public:
        static int foo() { return 10; }
        static int bar() { return 10; }
};

void quack(void)
{
    if (Moo::foo());
}
avj@stoat:~/cvise> ./cvise/build/clang_delta/clang_delta --query-instances=replace-function-def-with-decl test.cpp
Available transformation instances: 3
Available transformation instances:
 1 Moo::foo
 2 Moo::bar
 3 quack
```

**Comment by marxin** [comment]
> Sorry if this is a "dumb" question, but what's the status of making changes to `clang_delta` in the `cvise` repo?

Well, I basically take patches from the upstream repository and have some functionality on top of that.

> I am not sure how closely this repo tracks `creduce`, or if `cvise` is happy to have feature extensions to `clang_delta`.

You can suggest the change to upstream, but it's not responsive much :)

> 
> Why am I asking? Well, I am interested (in some circumstances) in running `clang_delta` to get the instance number _as well as the function name_ for a given transformation.
> 
> For some inputs, I know _a priori_ a function I want to _preserve_, but I want to replace every other function with the declaration. If `query-instances` could not only give the number of instances, but also the function _name_ for each instance, that would be awesome.

I see. Btw. it's a quite common requirement for anybody who is doing a reduction. I typically solve it easy by grepping in the interestingness test. May I ask you why do you need a more fine controll?

> 
> If I had this, I could do something like:
> 
> ```
> # query all of the instances in our original file
> clang_delta --query-instances=replace-function-def-with-decl test.cpp
> 
> # record the index of the function I care about; let's say it is index 3
> 
> # replace from the first to the second (leaving the third alone!)
> clang_delta --transformation=replace-function-def-with-decl --counter=1 --to-counter=2 test.cpp > intermediate.cpp
> 
> # re-query to get the remaining numbers
> clang_delta --query-instances=replace-function-def-with-decl intermediate.cpp
> 
> # validate that the original function should now be index 1
> 
> # replace from the second to the total number
> clang_delta --transformation=replace-function-def-with-decl --counter=2 --to-counter=99 intermediate.cpp > final.cpp
> 
> # the file final.cpp now *only* has a body for the function that was at index 3 in the original file
> ```
> 
> I've taken a look at this, and it seems that it is achievable by using `getQualifiedNameAsString` inside of `addOneFunctionDef`.

So you basically want to preserve a function given by name. That can be a different approach saying `clang_delta` `--preserve=myfunction`.

> 
> If I was to wrap this up as a feature request, is it something that `cvise` would be happy to accept?

Yes. I'm very open to a custom needs. But as mentioned, I would like to first know why exactly do you need the feature.

Cheers.

**Comment by andrewvaughanj** [comment]
We are using `cvise` to reduce instances that cause our tool to crash; for the most part, these crashes are 100% local to one routine/method; we still need types, etc. for the functions, but the bodies of the routines themselves are irrelevant (apart from the one routine we care about).

We actually have automation for a number of "pre-reduction" steps:

- Flatten system headers back-down (for our tool, bugs are always in "user" code, never system code)
- Remove comments (using `scc`)
- Remove line directives (using `grep -v`)

We've typically found it very effective if we are able to strip out the already "non interesting" routines, as they just bloat the instance, and `cvise` takes time reducing routines that are typically already redundant.

We actually wrote an external utility for this already, which works very well _for C_. However, for C++, and with inline methods, things don't work out. For C, we'd find the first and last curly, and then insert an `#if 0`/`#endif`. For inline class methods in C++, this is hard to do, because they are typically on one line.

> So you basically want to preserve a function given by name. That can be a different approach saying clang_delta --preserve=myfunction.

100%. The reason I *didn't* propose this is that I considered it a more intrusive change.

Does the above seem like a reasonable request? If so, I am happy to work on the `--preserve=myfunction` approach.

**Comment by andrewvaughanj** [comment]
Haven't opened the PR yet, but take a look at:

- https://github.com/marxin/cvise/compare/master...andrewvaughanj:preserve_routine?expand=1

If we're happy with this idea/approach, I can open-up a PR.

**Comment by marxin** [comment]
> We are using `cvise` to reduce instances that cause our tool to crash; for the most part, these crashes are 100% local to one routine/method; we still need types, etc. for the functions, but the bodies of the routines themselves are irrelevant (apart from the one routine we care about).

I see. May I ask what's the name of the tool you're trying to reduce test cases for?

> 
> We actually have automation for a number of "pre-reduction" steps:
> 
> * Flatten system headers back-down (for our tool, bugs are always in "user" code, never system code)

Do you mean that you remove some of the includes? If so, C-Vise has `cvise/passes/includes.py` pass.

> * Remove comments (using `scc`)

C-Vise has a special pass for that `cvise/passes/comments.py`. It should be pretty efficient?

> * Remove line directives (using `grep -v`)

Likewise here, take a look at `cvise/passes/line_markers.py`.

> 
> We've typically found it very effective if we are able to strip out the already "non interesting" routines, as they just bloat the instance, and `cvise` takes time reducing routines that are typically already redundant.

I see!

> 
> We actually wrote an external utility for this already, which works very well _for C_. However, for C++, and with inline methods, things don't work out. For C, we'd find the first and last curly, and then insert an `#if 0`/`#endif`. For inline class methods in C++, this is hard to do, because they are typically on one line.
> 
> > So you basically want to preserve a function given by name. That can be a different approach saying clang_delta --preserve=myfunction.
> 
> 100%. The reason I _didn't_ propose this is that I considered it a more intrusive change.

It will be simpler actually.

> 
> Does the above seem like a reasonable request? If so, I am happy to work on the `--preserve=myfunction` approach.

Looks quite good, I have some comments, but we can do it in a real PR.
We will need to add a `cvise` option which will be then passed to `clang_delta`, but I can help you with that.

**Comment by andrewvaughanj** [comment]
> I see. May I ask what's the name of the tool you're trying to reduce test cases for?
>

It is an "industrial-scale" test-case generator for C/C++ -- behaviourally, is very much something like [KLEE](https://klee.github.io) (i.e., a symbolic execution engine), but it shares no code or anything like that.

> Do you mean that you remove some of the includes? If so, C-Vise has cvise/passes/includes.py pass.
>

We start with a fully-expanded TU, but then use the line directives to "flatten" headers that come from system libraries. For example, we don't care about having a `#include <stdint.h>` in the final file, but we don't want `#include "some_magical_header.h"`. So we start with the fully-expanded TU, and collapse system headers back down, but leaving the user headers expanded.

> C-Vise has a special pass for that cvise/passes/comments.py. It should be pretty efficient?
> Likewise here, take a look at cvise/passes/line_markers.py.

I hadn't seen these -- I'll take a look!

> Looks quite good, I have some comments, but we can do it in a real PR.
>

PR opened!

**Comment by andrewvaughanj** [comment]
> It is an "industrial-scale" test-case generator for C/C++ -- behaviourally, is very much something like KLEE (i.e., a symbolic execution engine), but it shares no code or anything like that.
>

When the test-case generator crashes, it crashes _in a single routine_ (and it is mostly **not** intraprocedural). This means that, if we have a multi-MB TU, we can get significant reductions by removing the bodies of the remaining routines (and leaving only the routine that crashes the test-case generator).

**Comment by andrewvaughanj** [comment]
Closed in #44
